generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -167,10 +167,24 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
   }
 
   /**
-   * Function to generate an opaque zero or one by bitwise shifting left or right by a value
-   * between 0 and 8 bits, inclusive. For example, we can transform an integer i -> i >> 0 or
-   * i << 0. Generates an opaque value if the type of zero to be generated is an integer and the
-   * shading language in use supports bitwise operations.
+   * Function to generate an opaque zero or one by bitwise shifting left or right by an amount of
+   * bits dependent on which opaque is being generated.
+   * The minimum precision for a lowp integer in GLSL is 9 bits, with one of those bits reserved as
+   * a sign bit if the integer is signed. The OpenGL specification does not define behavior for
+   * shifting an integer beyond its maximum size in bits.
+   * This influences the process of generating opaque values such that:
+   *     If we're generating an opaque zero, we don't have to worry about losing bits, so our
+   *     maximum shift value is 8 bits.
+   *     If we're generating an opaque one, we have to make sure that we don't lose our 1 bit.
+   *     This limits our maximum shift value to 7 bits (more on how we shift an opaque one below).
+   * Possibilities for generating an opaque zero include:
+   *     Shifting an opaque zero by n bits: (opaque zero) >> n or (opaque zero) << n, where n is
+   *     an expression of a clamped value between 0 and 8:
+   *     n = clamp(fuzzedexpr, (opaque zero), identity(typeconstructor(8)).
+   * Possibilities for generating an opaque zero include:
+   *     Shifting an opaque one to the left by n bits, then shifting it to the right by n bits:
+   *     ((opaque one) << n) >> n, where n is an expression of a clamped value between 0 and 7,
+   *     inclusive: n = clamp(fuzzedexpr, (opaque zero), identity(typeconstructor(7)).
```

Comment:
The above is really excellent!

Reply:
Thanks, it took awhile for me to figure out how I wanted to phrase it.